[CLI] Include validation checks for Compute Resources with multiple instance types

EddyMM opened the pull request:
### Description of changes
- EFA Requirements:
    - Instance types should have the same EFA support status if EFA is enabled in the Compute Resource
    - If EFA is not enabled and one or more instance types support EFA, print a warning validation message
- Networking Requirements:
    - The lowest value for the MaximumNetworkInterfaceCards among the Instance Types is used. The user needs to be warned about this selection.
    - Compute Resources with multiple instance types with a PlacementGroup increases chances of an Insufficient Capacity Error. The user needs to be warned about this.
- CPU requirements:
    - Instance types should have the same number of CPUs or same number of Cores if Simultaneous Multithreading is disabled
- Accelerator requirements:
    - Instance types should have the same number of accelerators. The accelerators should also have the same manufacturer

### Tests
* Automated Tests to check validation messages for:
    * Instance Types should have the same number of CPUs
    * InstanceTypes should have the same number of cores if simultaneous multithreading is disabled
    * Instance Types should have the same number of Accelerators
    * Instance Types should have the same GPU manufacturer
    * Instance Types should have the same Accelerator Name (Inferentia)
    * If EFA is NOT enabled and one or more instance types supports EFA, a WARNING message should be printed
    * Instance Types should have the same EFA support status if EFA is enabled
    * Instance Types with varying Maximum NICs will have the smallest one used when setting the launch template
    * Using a placement group while having compute resources with multiple instance types increases the chances of getting an Insufficient Capacity Error

### References
* [describe-instance-types](https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-instance-types.html)

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

lukeseawalker commented on cli/tests/pcluster/validators/test_cluster_validators.py:
```diff
@@ -190,6 +191,62 @@ def test_schedulable_memory_validator(schedulable_memory, ec2memory, instance_ty
+@pytest.mark.parametrize(
+    "queue_name, compute_resource_name, instance_types_info, disable_simultaneous_multithreading, efa_enabled, "
+    "placement_group_enabled, expected_message",
+    [
```
add a test case where there is no failure, for both disable_simultaneous_multithreading choices

lukeseawalker commented on cli/src/pcluster/aws/aws_resources.py:
```diff
@@ -183,6 +183,36 @@ def gpu_count(self):
+            return list({gpu.get("Manufacturer") for gpu in gpu_info.get("Gpus", [])})
+        return []
+
+    def inference_accelerator_names(self):
```
`inference_accelerator_manufacturer` without the `s`. I do not expect an instance type have more than 1 inference accelerator of different manufacturers.

Thus, this should return a single value, not a list.

lukeseawalker commented on cli/src/pcluster/aws/aws_resources.py:
```diff
@@ -183,6 +183,36 @@ def gpu_count(self):
+        inference_accelerator_info = self.instance_type_data.get("InferenceAcceleratorInfo", None)
+
+        if inference_accelerator_info:
+            return list({accelerator.get("Name") for accelerator in inference_accelerator_info.get("Accelerators", [])})
```
`accelerator.get("Name")` -> `accelerator.get("Manufacturer")`

lukeseawalker commented on cli/src/pcluster/aws/aws_resources.py:
```diff
@@ -183,6 +183,36 @@ def gpu_count(self):
 
         return gpu_count
 
+    def gpu_manufacturers(self):
```
`gpu_manufacturer` without the `s`. I do not expect an instance type have more than 1 gpu of different manufacturers.

Thus, this should return a single value, not a list.

Can we move the warning "ParallelCluster currently does not offer native support for '%s' GPUs.  etc..." in this method. I think it's more appropriate having the check when checking the manufacturer rather when counting the GPUs

EddyMM replied:
> Thus, this should return a single value, not a list

👍, Just did a quick filter on EC2 console, all the instance types seem to return a single manufacturer.

> it's more appropriate having the check when checking the manufacturer

+1

lukeseawalker commented on cli/src/pcluster/aws/aws_resources.py:
```diff
@@ -183,6 +183,36 @@ def gpu_count(self):
+                if accelerator_name.lower() == "inferentia":
+                    accelerator_count += accelerator.get("Count", 0)
+                else:
+                    LOGGER.warning("ParallelCluster currently does not support %s accelerators.", accelerator_name)
```
I would rather move this on the `inference_accelerator_manufacturer` method. Then we can align the messaging similar to the one for the GPU case, i.e. "ParallelCluster currently does not offer native support for ..."

This should be a FailureLevel.WARNING, otherwise it only appears in the log

lukeseawalker commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -120,6 +121,67 @@ def validate_cpu_requirements(
+    ):
+        """Instance Types should have the same manufacturer type."""
+        unique_gpu_manufacturers = set()
+        for instance_type_info in instance_types_info.values():
```
this will be simplified if we consider only one manufacturer per instance type

lukeseawalker commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -120,6 +121,67 @@ def validate_cpu_requirements(
+        self.validate_size(
+            {instance_type_info.gpu_count() for instance_type_info in instance_types_info.values()},
+            1,
+            f"Instance types listed under Compute Resource {compute_resource_name} must have the same number of GPUs.",
```
Can we print the different values we have found for the count?

lukeseawalker commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -120,6 +121,67 @@ def validate_cpu_requirements(
+        self.validate_size(
+            {instance_type_info.inference_accelerator_count() for instance_type_info in instance_types_info.values()},
+            1,
+            f"Instance types listed under Compute Resource {compute_resource_name} must have the same number of "
```
Can we print the different values we have found for the count?

lukeseawalker commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -120,6 +121,67 @@ def validate_cpu_requirements(
+            unique_gpu_manufacturers,
+            1,
+            (
+                f"Instance types listed under Compute Resource {compute_resource_name} must have the same GPU "
```
Can we print the different values we have found for the manufacturers?

lukeseawalker commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -120,6 +121,67 @@ def validate_cpu_requirements(
+            unique_accelerator_names,
+            1,
+            (
+                f"Instance types listed under Compute Resource {compute_resource_name} must have the same inference "
```
Can we print the different values we have found for the manufacturers?

lukeseawalker commented on cli/tests/pcluster/validators/test_cluster_validators.py:
```diff
@@ -225,6 +225,138 @@ def test_schedulable_memory_validator(schedulable_memory, ec2memory, instance_ty
             "Instance types listed under Compute Resource TestComputeResource must have the same number of CPU "
             "cores when Simultaneous Multithreading is disabled.",
         ),
+        # Instance Types should have the same number of GPUs
```
Provide all the tests where there is no failure, e.g. same count, same manufacturer are respected

lukeseawalker commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -73,6 +73,54 @@
+        """
+        if disable_simultaneous_multithreading:
+            self.validate_size(
+                {instance_type_info.cores_count() for instance_type_info in instance_types_info.values()},
```
code it's not optimal since the check can fail as soon as there is a mismatch, without having to traverse all the list. But deviation isn't much if the list isn't big. Let's follow up on this

lukeseawalker commented on cli/src/pcluster/aws/aws_resources.py:
```diff
@@ -253,7 +252,7 @@ def supported_architecture(self):
     def is_efa_supported(self):
         """Check whether EFA is supported."""
-        return self.instance_type_data.get("NetworkInfo").get("EfaSupported")
+        return self.instance_type_data.get("NetworkInfo", {}).get("EfaSupported")
```
what about returning a default for "EfaSupported"? i.e. `.get("EfaSupported", false)`

lukeseawalker commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -182,6 +183,63 @@ def validate_accelerator_requirements(
+
+            # If all the instance types have EFA support, `instance_types_without_efa_support` should be empty
+            # --> No failure expected
+            # If all the instance types DO NOT have EFA support, `instance_types_without_efa_support` should be the same
```
> If all the instance types DO NOT have EFA support, `instance_types_without_efa_support` should be the same as `all_instance_types` (Set difference of 0) --> No failure expected

I think this should be a failure because EFA is enabled on LT and NO instance will be launched, since none of them support EFA

lukeseawalker commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -182,6 +183,63 @@ def validate_accelerator_requirements(
+            if instance_types_without_efa_support and len(all_instance_types - instance_types_without_efa_support) > 0:
+                self._add_failure(
+                    (
+                        "Instance types ({0}) in Compute Resource {1} do not support EFA.".format(
```
`types` -> `type(s)`

lukeseawalker replied:
what about adding the reason of the failure here? Something like "and cannot be launched because you have enabled EFA [in CR configuration]"

lukeseawalker commented on cli/tests/pcluster/validators/test_cluster_validators.py:
```diff
@@ -357,6 +357,90 @@ def test_schedulable_memory_validator(schedulable_memory, ec2memory, instance_ty
             "Instance types listed under Compute Resource TestComputeResource must have the same inference "
             "accelerator manufacturer",
         ),
+        # Instance Types should have the same EFA support status if EFA is enabled
```
add the test cases when there is no failure (e.g EFA enabled and all instances support EFA, EFA disabled and all instances don't support EFA)

lukeseawalker commented on cli/tests/pcluster/validators/test_cluster_validators.py:
```diff
@@ -441,6 +441,43 @@ def test_schedulable_memory_validator(schedulable_memory, ec2memory, instance_ty
             "update the cluster's configuration to enable EFA ("
             "https://docs.aws.amazon.com/parallelcluster/latest/ug/efa-v3.html).",
         ),
+        # Instance Types with varying Maximum NICs will have the smallest one used when setting the launch template
```
add the test cases for when there is no WARNING

lukeseawalker commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -278,6 +281,18 @@ def validate_networking_requirements(
+        """On-demand Capacity type only supports "lowest-price" allocation strategy."""
+        if (
+            capacity_type == cluster_config.CapacityType.ONDEMAND
+            and allocation_strategy == cluster_config.AllocationStrategy.CAPACITY_OPTIMIZED
```
change to `allocation_strategy` != `LOWEST_PRICE` so that if a new allocation strategy is added and not supported by ONDEMAND, the check will be still valid

lukeseawalker commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -278,6 +281,18 @@ def validate_networking_requirements(
+        ):
+            self._add_failure(
+                f"Compute Resource {compute_resource_name} is using an OnDemand CapacityType. OnDemand CapacityType "
+                "can only use 'lowest-price' allocation strategy.",
```
use the enum instead of the 'lowest-price' string, so that if value changes there is only one place where it has to be changed and message will be still valid

lukeseawalker commented on cli/tests/pcluster/validators/test_cluster_validators.py:
```diff
@@ -478,10 +500,26 @@ def test_schedulable_memory_validator(schedulable_memory, ec2memory, instance_ty
             "use of multiple instance types for Compute Resource: TestComputeResource ("
             "https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/placement-groups.html#placement-groups-cluster).",
         ),
+        # OnDemand Capacity type only supports "lowest-price" allocation strategy
```
add test cases when there is no failure

lukeseawalker commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -73,6 +73,246 @@
+        placement_group_enabled: bool,
+        memory_scheduling_enabled: bool,
+    ):
+        self.validate_cpu_requirements(compute_resource_name, instance_types_info, disable_simultaneous_multithreading)
```
I'm worried about putting all the checks in a single validator. This because if user wants to disable validator to bypass cpu check, then all the other checks will be bypassed as well. I'm rather on the opinion that the check needs to be disabled on demand, one by one. Let's follow up on this.

EddyMM committed:
[CLI] Include CPU validation checks for Compute Resources with multiple instance types

- CPU requirements:
    - Instance types should have the same number of CPUs or same number of Cores if Simultaneous Multithreading is disabled

Signed-off-by: Eddy Mwiti <eddmwiti@amazon.com>

EddyMM committed:
[CLI] Include Accelerator validation checks for Compute Resources with multiple instance types

- Accelerator requirements:
    - Instance types should have the same number of accelerators. The accelerators should also have the same manufacturer

Signed-off-by: Eddy Mwiti <eddmwiti@amazon.com>

EddyMM committed:
[CLI] Include EFA validation checks for Compute Resources with multiple instance types

- EFA Requirements:
    - Instance types should have the same EFA support status if EFA is enabled in the Compute Resource
    - If EFA is not enabled and one or more instance types support EFA, print a warning validation message

Signed-off-by: Eddy Mwiti <eddmwiti@amazon.com>

EddyMM committed:
[CLI] Include Networking validation checks for Compute Resources with multiple instance types

- Networking Requirements:
    - The lowest value for the MaximumNetworkInterfaceCards among the Instance Types is used. The user needs to be warned about this selection.
    - Compute Resources with multiple instance types with a PlacementGroup increases chances of an Insufficient Capacity Error. The user needs to be warned about this.

Signed-off-by: Eddy Mwiti <eddmwiti@amazon.com>

EddyMM committed:
[CLI] Include Allocation Strategy validation checks for Compute Resources with multiple instance types

- AllocationStrategy/CapacityType Requirements:
    - OnDemand Capacity type only supports "lowest-price" allocation strategy
    - Spot Capacity type supports both "lowest-price" and "capacity-optimized" allocation strategy

Signed-off-by: Eddy Mwiti <eddmwiti@amazon.com>

enrico-usai commented on cli/src/pcluster/aws/aws_resources.py:
```diff
@@ -164,28 +164,85 @@ def __init__(self, instance_type_data):
+
+        # Only one GPU manufacturer is associated with each Instance Type's GPU
+        manufacturer = gpu_manufacturers[0] if gpu_manufacturers else ""
+        if manufacturer.upper() != "NVIDIA":
```
This warning message will be printed even if manufacturer is empty. Is this expected?

enrico-usai commented on cli/src/pcluster/aws/aws_resources.py:
```diff
@@ -164,28 +164,85 @@ def __init__(self, instance_type_data):
+
+    def inference_accelerator_manufacturer(self) -> str:
+        """Return the list of Inference Accelerator Manufacturers supported by this instance type."""
+        inference_accelerator_info = self.instance_type_data.get("InferenceAcceleratorInfo", None)
```
to save an if/else you could do:
```python
inference_accelerator_info = self.instance_type_data.get("InferenceAcceleratorInfo", {})
...
inference_accelerator_manufacturers = (
            list(
                {
                    accelerator.get("Manufacturer", "")
                    for accelerator in inference_accelerator_info.get("Accelerators", [])
                }
            )
        )
```

enrico-usai commented on cli/src/pcluster/aws/aws_resources.py:
```diff
@@ -164,28 +164,85 @@ def __init__(self, instance_type_data):
+        """Return the list of GPU manufacturers supported by this instance type."""
+        gpu_info = self.instance_type_data.get("GpuInfo", None)
+
+        gpu_manufacturers = list({gpu.get("Manufacturer", "") for gpu in gpu_info.get("Gpus", [])}) if gpu_info else []
```
to save an if/else you can do:
```python
gpu_info = self.instance_type_data.get("GpuInfo", {})

gpu_manufacturers = list({gpu.get("Manufacturer", "") for gpu in gpu_info.get("Gpus", [])})
```

enrico-usai commented on cli/src/pcluster/aws/aws_resources.py:
```diff
@@ -164,28 +164,85 @@ def __init__(self, instance_type_data):
+        # Only one accelerator manufacturer is associated with each Instance Type's accelerator
+
+        accelerator_manufacturer = inference_accelerator_manufacturers[0] if inference_accelerator_manufacturers else ""
+        if accelerator_manufacturer.upper() != "AWS":
```
same here, we'll print a warning with an empty accelerator_manufacturer, is this expected?

lukeseawalker replied:
I would say yes, since there is an accelerator data but the manufacturer is missing

EddyMM replied:
Good point. I think logging that only 'AWS' manufacturer is supported, then show the `accelerator_info`.
This will be enough to communicate why the accelerator wasn't considered.

enrico-usai commented on cli/src/pcluster/aws/aws_resources.py:
```diff
@@ -164,28 +164,85 @@ def __init__(self, instance_type_data):
+        return manufacturer
+
+    def inference_accelerator_manufacturer(self) -> str:
+        """Return the list of Inference Accelerator Manufacturers supported by this instance type."""
```
wrong comment: this is not returning a list.

enrico-usai commented on cli/src/pcluster/aws/aws_resources.py:
```diff
@@ -164,28 +164,85 @@ def __init__(self, instance_type_data):
         return gpu_count
 
+    def gpu_manufacturer(self) -> str:
+        """Return the list of GPU manufacturers supported by this instance type."""
```
wrong comment: this is not returning a list.

enrico-usai commented on cli/src/pcluster/aws/aws_resources.py:
```diff
@@ -164,28 +164,85 @@ def __init__(self, instance_type_data):
+
+    def inference_accelerator_count(self):
+        """Return the total number of Inference Accelerators associated with this instance type."""
+        inference_accelerator_info = self.instance_type_data.get("InferenceAcceleratorInfo", None)
```
same here, you can save the if by doing:
```python
inference_accelerator_info = self.instance_type_data.get("InferenceAcceleratorInfo", {})
``` 

enrico-usai commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1806,21 +1812,32 @@ def queue_ami(self):
         if any(isinstance(compute_resource, SlurmFlexibleComputeResource) for compute_resource in compute_resources):
-            self.allocation_strategy = allocation_strategy
+            self.allocation_strategy = (
+                AllocationStrategy[allocation_strategy.replace("-", "_").upper()]
```
you can use `to_camel_case` utility function in `utils.py`

EddyMM replied:
Thanks, I think `to_snake_case` is the right utility function.

enrico-usai commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -2462,8 +2479,32 @@ def _register_validators(self):
+                        memory_scheduling_enabled=self.scheduling.settings.enable_memory_based_scheduling,
+                    )
+                    flexible_instance_types_validators = [
+                        InstanceTypesListCPUValidator,
```
we should rename all of them to be: `InstanceTypeList`, rather than `InstanceTypesList` to be consistent with the configuration parameter name.

enrico-usai commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -1151,3 +1151,258 @@ def _validate(self, scheduler):
+        """
+        if disable_simultaneous_multithreading:
+            self.validate_property_homogeneity(
+                instance_type_info_list=list(instance_types_info.values()),
```
to avoid repetition and simplify the code you could:
1. accept `instance_types_info` in  `validate_property_homogeneity` and make the get of values and conversion to a list within the `validate_property_homogeneity` or alternatively you can define a method that does it in the `_FlexibleInstanceTypesValidatorMixin` and just call it every time you need to perform the conversion
1. pass just a piece of the message and add the initial part that is it in common  within the `validate_property_homogeneity` (I mean: `Instance types listed under Compute Resource`) 
1. avoid to pass `failure_level given` that it's always set to ERROR and just set it internally
1. (optional) instead of passing a callback function for the `property callback` you could simply pass the method/property name (i.e. `"vcpus_count"`, `"cores_count"`, ...).

EddyMM replied:
1. 👍
2. I'm giving the caller of `validate_property_homogeneity` flexibility to use the mismatched values for now, since the wording hasn't been decided on yet.
3. 👍  I can set `ERROR` as the default argument for the `failure_level` parameter and avoid passing it explicitly.
4. I chose a callback to give the caller the flexibility to process the `instance_type_info` object

enrico-usai commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -1151,3 +1151,258 @@ def _validate(self, scheduler):
+            )
+
+
+class InstanceTypesListAcceleratorsValidator(Validator, _FlexibleInstanceTypesValidatorMixin):
```
InstanceTypeList (without the s), everywhere.

enrico-usai commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -1151,3 +1151,258 @@ def _validate(self, scheduler):
+            if instance_types_without_efa_support and len(all_instance_types - instance_types_without_efa_support) > 0:
+                self._add_failure(
+                    (
+                        "Instance type(s) ({0}) in Compute Resource {1} do not support EFA and cannot be launched "
```
What about: 
`"Instance type(s) ({0}) do not support EFA and cannot be launched when EFA is enabled in Compute Resource: {1}."`

enrico-usai commented on cli/tests/pcluster/validators/test_cluster_validators.py:
```diff
@@ -1162,3 +1171,568 @@ def test_root_volume_size_validator(mocker, root_volume_size, ami_size, expected
+                        "GpuInfo": {
+                            "Gpus": [
+                                {
+                                    "Name": "A10G",
```
To simplify test code you could avoid to define items in dictionary that are not used by the validator. (maybe it's already so).

enrico-usai commented on cli/tests/pcluster/validators/test_cluster_validators.py:
```diff
@@ -1162,3 +1171,568 @@ def test_root_volume_size_validator(mocker, root_volume_size, ami_size, expected
+                    }
+                ),
+            },
+            "Instance types listed under Compute Resource TestComputeResource must have the same GPU manufacturer ({"
```
Up to you: You could add `.*` in place of the common part of the message `Instance types listed under Compute Resource`

EddyMM replied:
Thanks, I'll use `.*` for messages in the logs. For the validator messages I'd prefer matching the exact text since they're customer facing.

enrico-usai commented on cli/tests/pcluster/validators/test_cluster_validators.py:
```diff
@@ -1162,3 +1171,568 @@ def test_root_volume_size_validator(mocker, root_volume_size, ami_size, expected
     assert_failure_messages(actual_failures, expected_message)
+
+
+@pytest.mark.parametrize(
```
I think it's worth to move your validator into a separate module. 
So move them from `cluster_validators` to `instance_type_list_validators` and do the same for the tests: from `test_cluster_validators` to `test_instance_type_list_validators`.

enrico-usai commented on cli/tests/pcluster/validators/test_cluster_validators.py:
```diff
@@ -1162,3 +1171,568 @@ def test_root_volume_size_validator(mocker, root_volume_size, ami_size, expected
+            {
+                "t2.micro": InstanceTypeInfo(
+                    {
+                        "VCpuInfo": {"DefaultVCpus": 2, "DefaultCores": 2},
```
For example, do we really need vcpu info here to test this validator?

EddyMM replied:
The `vCPUInfo` can be removed for some of the validators.

enrico-usai commented on cli/tests/pcluster/validators/test_cluster_validators.py:
```diff
@@ -1162,3 +1171,568 @@ def test_root_volume_size_validator(mocker, root_volume_size, ami_size, expected
+                "c5n.18xlarge": InstanceTypeInfo(
+                    {
+                        "VCpuInfo": {"DefaultVCpus": 2, "DefaultCores": 2},
+                        "NetworkInfo": {"EfaSupported": False},
```
I'd use t3.micro and t2.micro in this example, to be consistent with EfaSupported False value.

enrico-usai commented on cli/tests/pcluster/validators/test_cluster_validators.py:
```diff
@@ -1162,3 +1171,568 @@ def test_root_volume_size_validator(mocker, root_volume_size, ami_size, expected
+                ),
+                "t3.micro": InstanceTypeInfo(
+                    {
+                        "VCpuInfo": {"DefaultVCpus": 4, "DefaultCores": 2},
```
same here, do we really need vcpu info to test this validator?

enrico-usai commented on cli/tests/pcluster/validators/test_cluster_validators.py:
```diff
@@ -1162,3 +1171,568 @@ def test_root_volume_size_validator(mocker, root_volume_size, ami_size, expected
+@pytest.mark.parametrize(
+    "compute_resource_name, instance_types_info, memory_scheduling_enabled, expected_message",
+    [
+        # Memory-based scheduling is supported for Compute Resource that use either 'InstanceType' or have a sinlge
```
typo: single

lukeseawalker commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -82,6 +82,7 @@
     HeadNodeLaunchTemplateValidator,
     HostedZoneValidator,
     InstanceArchitectureCompatibilityValidator,
+    InstanceTypesListCPUValidator,
```
InstanceTypesListCPUValidator -> InstanceTypeListCPUValidator

lukeseawalker commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -1272,3 +1254,76 @@ def _validate(
+
+            # If all the instance types have EFA support, `instance_types_without_efa_support` should be empty
+            # --> No failure expected
+            # If all the instance types DO NOT have EFA support, `instance_types_without_efa_support` should be the same
```
> If all the instance types DO NOT have EFA support, instance_types_without_efa_support should be the same as all_instance_types (Set difference of 0) --> No failure expected

Please fix this: this should be a failure because EFA is enabled on LT and NO instance will be launched, since none of them support EFA

lukeseawalker commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -1272,3 +1254,76 @@ def _validate(
+                    ),
+                    FailureLevel.ERROR,
+                )
+            if len(all_instance_types - instance_types_without_efa_support) == 0:
```
I believe the two if can be simplified into just one `instance_types_without_efa_support > 0`.

EddyMM replied:
I agree, we can use the same messaging as well.

EddyMM committed:
[CLI] Include Memory Scheduling validation checks for Compute Resources with multiple instance types

- Memory Scheduling Requirements:
    - Memory-based scheduling is supported for Compute Resource that use either 'InstanceType' or have a single instance type under 'InstanceTypeList'

Signed-off-by: Eddy Mwiti <eddmwiti@amazon.com>

enrico-usai reviewed (approved):
Great job!
